### ISSUE TITLE
Pytorch 1.7.0 RuntimeError

### ISSUE THREAD

**Issue by Holldean** [shenweichen/DeepCTR-Torch#130]
**Describe the bug(问题描述)**
Install deepctr-torch using `pip install -U deepctr-torch`, run criteo example but get an error: 
```RuntimeError: Output 0 of UnbindBackward is a view and its base or another view of its base has been modified inplace. This view is the output of a function that returns multiple views. Such functions do not allow the output views to be modified inplace. You should replace the inplace operation by an out-of-place one.```

**To Reproduce(复现步骤)**
1. conda create -n deepctr-torch
2. conda activate deepctr-torch
3. pip install -U deepctr-torch
4. git clone https://github.com/shenweichen/DeepCTR-Torch.git
5. cd ./DeepCTR-Torch/examples
6. python run_classification_criteo.py
7. get the RuntimeError

**Operating environment(运行环境):**
 - python version [3.8.3]
 - torch version [1.7.0]
 - deepctr-torch version [0.2.3]
 - deepctr version [0.8.2]

**Additional context**
Roll back pytorch to 1.6.0 and it runs normally.

**Comment by zanshuxun** [comment]
### 1.Cause
This is because pytorch 1.7 changes Tensor iteration behavior. When `weight_list` is a tensor or a Parameter, `list(weight_list)` returns list of **UnbindBackward** tensors, which further lead to inplace operation error.
In previous versions (1.2-1.6), `list(weight_list)` returns list of **SelectBackward** tensors and it works.
Related pytorch issue can be seen here: https://github.com/pytorch/pytorch/issues/47899

Location:

In DeepCTR-Torch, this error happens in:
https://github.com/shenweichen/DeepCTR-Torch/blob/bc881dcd417fec64f840b0cacce124bc86b3687c/deepctr_torch/models/deepfm.py#L63
https://github.com/shenweichen/DeepCTR-Torch/blob/bc881dcd417fec64f840b0cacce124bc86b3687c/deepctr_torch/models/basemodel.py#L371-L372



### 2.Solution
We fix this bug by simply remove `list(weight_list)` operation for <class 'torch.nn.parameter.Parameter'>  input in https://github.com/shenweichen/DeepCTR-Torch/commit/a9402a474481fd186cb8428d6df9aadc2b3c69dd.

Take DeepFM as example, `add_regularization_weight()` are called 4 times:
| file         | input (weight_list)                                          | type                                   |
| ------------ | ------------------------------------------------------------ | -------------------------------------- |
| basemodel.py | self.embedding_dict.parameters()                             | <class 'generator'>                    |
| basemodel.py | self.linear_model.parameters()                               | <class 'generator'>                    |
| deepfm.py    | filter(lambda x: 'weight' in x[0] and 'bn' not in x[0], self.dnn.named_parameters()) | <class 'filter'>                       |
| deepfm.py    | self.dnn_linear.weight                                       | <class 'torch.nn.parameter.Parameter'> |
 
`list(weight_list)` are used to convert generator and filter to a list of tensors. For a Parameter, we can directly compute its norm and there is no need to convert it to a list.






### 3.Reproduction
If you are still interested, I simplify this problem and reproduce it in the codes below (works in previous versions but fails in 1.7)
```python
import torch

x = torch.tensor([[0.5, 0.2, 0.3, 0.8]], requires_grad=True)
w = torch.tensor([[0.2, 0.5, 0.1, 0.5]], requires_grad=True)
y_true = torch.tensor([1])

# In 1.7, `list(w)` returns list of UnbindBackward tensors, which further leads to inplace operation error.
# In previous versions, `list(w)` returns list of SelectBackward tensors and it works.
weight_list = list(w)
print('weight_list',weight_list) 

for i in range(2):
    l2_norm = torch.norm(weight_list[0], 2)
    y = w.mm(x.T)

    # mse with l2 regularization
    loss = pow(y - y_true, 2) + 0.5 * pow(l2_norm, 2)
    loss.backward()

    # simulate optimizer.step()
    with torch.no_grad():
        w.add_(w.grad, alpha=1e-3)

```

output:
**torch 1.7:** 
```
weight_list [tensor([0.2000, 0.5000, 0.1000, 0.5000], grad_fn=<UnbindBackward>)]
...
RuntimeError: Output 0 of UnbindBackward is a view and its base or another view of its base has been modified inplace. This view is the output of a function that returns multiple views. Such functions do not allow the output views to be modified inplace. You should replace the inplace operation by an out-of-place one.

Process finished with exit code 1
```

**previous versions:** 
```
weight_list [tensor([0.2000, 0.5000, 0.1000, 0.5000], grad_fn=<SelectBackward>)]

Process finished with exit code 0
```

Except the solution above, if it's necessary to so such operation, using indices works:
```python
# weight_list = list(w)
weight_list = [w[i] for i in range(len(w))]
```

**Comment by zanshuxun** [comment]
Details can be found in https://github.com/pytorch/pytorch/issues/47899 and the release note of pytorch 1.7 [here](https://github.com/pytorch/pytorch/releases/tag/v1.7.0).
### Use `torch.unbind` instead of a for loop
![image](https://user-images.githubusercontent.com/28911886/99338620-82028580-28bf-11eb-9beb-416562991103.png)

**Comment by LeeTsinghua** [comment]
Thanks

**Comment by zanshuxun** [comment]
Hi, we have solved this issue in [v0.2.4](https://github.com/shenweichen/DeepCTR-Torch/releases/tag/v0.2.4), please use `pip install -U deepctr-torch` to upgrade.

**Comment by Holldean** [comment]
> Hi, we have solved this issue in [v0.2.4](https://github.com/shenweichen/DeepCTR-Torch/releases/tag/v0.2.4), please use `pip install -U deepctr-torch` to upgrade.

Thanks, I will try it later!
